### PR TITLE
⌚️ Add forum watchlist

### PR DESCRIPTION
**Pull request by p-sad**
See #1057 
Also updated `useLocalStorage` a bit. Item type can now be any type (because there's `JSON.stringify`/`parse` in there anyway) and the returned type is now `<T | undefined>` (TS didn't notice this, but `window.localStorage.getItem()` returns undefined if the key isn't there, and it's better to state that explicitly so it doesn't surprise us later)

### CODE REVIEW

File: packages/ui/src/common/hooks/useLocalStorage.ts [hunk 1]
```diff
@@ -23,8 +23,8 @@ function setItem(key: string, value: any) {
   }
 }
 
-export const useLocalStorage = <T extends string>(key: string) => {
-  const [value, setValue] = useState<T>(() => getItem(key))
+export const useLocalStorage = <T>(key: string) => {
+  const [value, setValue] = useState<T | undefined>(() => getItem(key))
```
Comment (p-sad) [inline packages/ui/src/common/hooks/useLocalStorage.ts:27]: Just realized there is no test for this behavior, will add

File: packages/ui/src/forum/components/threads/ThreadBrowser.tsx [hunk 2]
```diff
@@ -44,10 +44,16 @@ export const ThreadBrowser = ({
           {label} {!!totalCount && <CountBadge count={totalCount} />}
         </Label>
         <ButtonsGroup>
-          <ButtonGhost size="small" square onClick={onPrevClick} disabled={currentPage <= 1}>
+          <ButtonGhost size="small" square onClick={onPrevClick} disabled={currentPage <= 1} title="Browse left">
```
Comment (jodator) [inline packages/ui/src/forum/components/threads/ThreadBrowser.tsx:47]:
```suggestion
          <ButtonGhost size="small" square onClick={onPrevClick} disabled={currentPage <= 1} title="Browse previous">
```

File: packages/ui/src/forum/components/threads/ThreadBrowser.tsx [hunk 3]
```diff
@@ -44,10 +44,16 @@ export const ThreadBrowser = ({
...
             <Arrow direction="left" />
           </ButtonGhost>
-          <ButtonGhost size="small" square onClick={onNextClick} disabled={currentPage >= (pageCount ?? 1)}>
+          <ButtonGhost
+            size="small"
+            square
+            onClick={onNextClick}
+            disabled={currentPage >= (pageCount ?? 1)}
+            title="Browse right"
```
Comment (jodator) [inline packages/ui/src/forum/components/threads/ThreadBrowser.tsx:55]:
```suggestion
            title="Browse next"
```

File: packages/ui/src/common/hooks/useLocalStorage.ts (hunk 1 above)
Comment (jodator) [reply packages/ui/src/common/hooks/useLocalStorage.ts:27]: :eyes: AFAICS there's none :wink:

File: packages/ui/test/forum/components/WatchlistButton.test.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    renderComponent()
+    const button = await getButton('Watch thread')
+    act(() => {
+      fireEvent.click(button)
+    })
+    const watchlist = window.localStorage.getItem(FORUM_WATCHLIST)
+    expect(JSON.parse(watchlist!)).toEqual(['2', '1'])
+  })
+
+  it('Add then remove thread', async () => {
```
Comment (jodator) [inline packages/ui/test/forum/components/WatchlistButton.test.tsx:35]:
```suggestion
  it('Remove thread', async () => {
```

File: packages/ui/test/forum/components/WatchlistButton.test.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import { FORUM_WATCHLIST } from '@/forum/constant'
+
+import { getButton } from '../../_helpers/getButton'
+
+describe('UI: WatchlistButton', () => {
+  beforeEach(() => {
+    window.localStorage.clear()
+  })
+
+  it('Create a watchlist if there is none', async () => {
```
Comment (jodator) [inline packages/ui/test/forum/components/WatchlistButton.test.tsx:14]:
```suggestion
  it('Create a watchlist', async () => {
```

Comment (jodator) [reply packages/ui/test/forum/components/WatchlistButton.test.tsx:14]: I'd make the test names shorter

File: packages/ui/src/forum/components/Thread/WatchlistButton.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+interface Props {
+  threadId: string
+}
+
+export const WatchlistButton = ({ threadId }: Props) => {
+  const [watchlist, setWatchlist] = useLocalStorage<string[]>(FORUM_WATCHLIST)
+  const addToWatchlist = useCallback(() => setWatchlist(watchlist ? [...watchlist, threadId] : [threadId]), [watchlist])
+  const removeFromWatchlist = useCallback(() => setWatchlist(watchlist?.filter((id) => id !== threadId)), [watchlist])
+
+  const isThreadWatchlisted = watchlist?.includes(threadId)
```
Comment (jodator) [inline packages/ui/src/forum/components/Thread/WatchlistButton.tsx:17]:
```suggestion
  const isOnWatchlist = watchlist?.includes(threadId)
```

WDYT? I found that easier to read.

**Review by jodator — APPROVED** [review]
Nice. I've added some minor comments for language, but overall I had nothing to add there.

You can merge after those fixes.

File: packages/ui/src/forum/components/threads/ThreadBrowser.tsx (hunk 2 above)
Comment (jodator) [reply packages/ui/src/forum/components/threads/ThreadBrowser.tsx:47]: I forgot to write: Left/right is not accurate. Previous/Next is a bit better + it is safer for RTL languages as well ;)

File: packages/ui/test/forum/components/WatchlistThreadsBrowser.test.tsx [hunk 7]
```diff
@@ -14,6 +14,9 @@ import { setupMockServer } from '../../_mocks/server'
 describe('UI: WatchlistThreadsBrowser', () => {
   const server = setupMockServer({ noCleanupAfterEach: true })
 
+  const matchHeader = (regex: RegExp) => (content: string, element: Element | null) =>
+    element?.tagName.toLowerCase() == 'h5' && regex.test(content)
```
Comment (jodator) [inline packages/ui/test/forum/components/WatchlistThreadsBrowser.test.tsx:18]: Did you saw the [`selector` option for `findByText()`](https://testing-library.com/docs/queries/bytext#selector)? Shouldn't that do the trick?

Comment (p-sad) [reply packages/ui/test/forum/components/WatchlistThreadsBrowser.test.tsx]: It does, must've missed it

**Pull request merged**